generate_sets.py

```diff
@@ -0,0 +1,78 @@
+# -*- coding: utf-8 -*-
+"""Generate Sets of Training Data TextLine + Image Pairs"""
+
+import argparse
+import os
+
+from sets.training_sets import (
+    TrainingSets,
+    DEFAULT_MIN_CHARS
+)
+
+
+########
+# MAIN #
+########
+APP_ARGUMENTS = argparse.ArgumentParser()
+APP_ARGUMENTS.add_argument(
+    "-d",
+    "--data",
+    required=True,
+    help="path alto|page file")
+APP_ARGUMENTS.add_argument(
+    "-i",
+    "--image",
+    required=True,
+    help="path image file tif|jpg|png")
+APP_ARGUMENTS.add_argument(
+    "-o",
+    "--output",
+    required=False,
+    help="path output data")
+APP_ARGUMENTS.add_argument(
+    "-m",
+    "--minchars",
+    required=False,
+    help="Minimum chars for a line")
```

Comment:
Historical note: Originates from newspaper digitalization project, where a common text line (no adds) has usually more than 20 chars. In fact, we only took care for lines with at least 32 chars, since I thought these lines were more valuable for training than shorter lines because of having more characters to learn.
But personally I'm totally free about this, so what about, say, 4 chars?

Reply:
It makes sense to skip short lines for training, but the fact that there is a minimum of chars and what that minimum is should be clearly communicated to the user, so he/she isn't surprised why some lines are skipped.

And yes, probably something low like 4 (documented in `./generate_sets.py --help` would be best IMHO.

Reply:
It probably makes sense not to skip short lines for training. Tesseract was trained only with artifical long lines initially, and the standard models have problems with short lines (typically the page numbers, but also short lines ending paragraphs). We know that there exist valid lines with only a single character for page numbers (1 … 9, a … z, A … Z). Why should be skip lines with one or two characters as long as they are valid?

Reply:
This originates from the decision to prefer longer lines because they provide more characters. I thought more characters means more training character training material, and more material increases pattern recognition accuracy. But this doesn't pay much attention to a characters' context. In newspapers advertises I've seen many lines that are way shorter than 8 chars, only containing abbreviations and alike. Maybe focusing on "regular article lines" is another reason why Tesseract (4.1.1) usually performs rather worse in this realm, compared to single-column-liners. 

@stweil Do you suggest to turn the minchars arg into being completely optional, or to set the default value to "1", to skip lines that only contain non-printable characters?

Reply:
Setting minchars to 1 sounds reasonable. I cannot imagine how a line which only contains non-printable characters would look like.

Reply:
I agree with Stefan we should make minchars optional and try to make Tesseract learn short lines well. Not sure how the LSTM implementation here unrolls, but short lines should create fewer weight updates, so characters would still contribute "democratically" – just that there's more incentive to get a better transition from the initial state.